### PR TITLE
Allow public use statements to reexport for qualified access

### LINKED ISSUE #15282: Using 'use' to re-export
`use` statements don't currently allow providing an alternative path to a symbol for qualified access. This means that `use` wouldn't have the expected effect when used for re-exporting if the module is `import`d or `use only`d. Note though that these patterns are expected to be possible to write with import.

``` chapel
module A {
  var x: int;
}

module B {
  public use A;
}

module Program {
  import B; // or use B only;

  proc main() {
    //writeln(B.x); // does not work today
    //writeln(B.A.x); // does not work today
  }
}
```

I would expect the `B.x` or `B.A.x` calls above to function.

### PR DESCRIPTION
**Pull request by lydia-duncan**
Import statements introduced the concept of reexporting, where a public
import statement would cause the module or symbols imported to appear
as though they were defined within the module with the public import.  Use
statements already partially appeared this way (symbols made publicly
available for unqualified access were also available for unqualified access to
modules that used the module with the public use), but qualified access was
not made possible in a similar way.  This PR seeks to enable reexporting in
use statements to extend to qualified access.

Resolves #15282

<details>

Makes skipSymbolSearch a pure virtual method on the VisibilityStmt type so
that code can be reused instead of duplicated for UseStmts.  As a result, the
ImportStmt version also got marked as const.

Adds a boolean argument to lookupPublicUnqualAccessSyms indicating whether
we should follow UseStmts for reexporting in the current context (set to false
by default and true when used for qualified access).  This is because that
function is called both for qualified and unqualified access, and we don't want
unqualified access involving use statements to be changed.

The rest of the compiler changes are just reusing code instead of duplicating it,
and renaming some symbols to match the extended functionality.
</details>

Test changes:
- Extends and resolves the future so that it also demonstrates that the modules
being publicly used can be treated as those they are submodules of the module
with the public use
- Adds a copy of the future that imports the module with the public uses instead
of using that module.

Full paratest with futures passed (though it did take longer than I expected - this
may be because the change slows down compilation, or it may be due to running
testing first thing in the morning when not as many nodes are available)

### CODE REVIEW

File: compiler/passes/ResolveScope.cpp [hunk 1]
```diff
@@ -980,11 +980,11 @@ Symbol* ResolveScope::lookupPublicImports(const char* name) const {
   Symbol *retval = NULL;
```
Comment (e-kayrakli) [inline compiler/passes/ResolveScope.cpp:980]: Can't comment on that specific line: I think this function should be renamed to `lookupPublicVisStmts` or `lookupPublicUsesAndImports`

File: compiler/passes/ResolveScope.cpp [hunk 2]
```diff
@@ -980,11 +980,11 @@ Symbol* ResolveScope::lookupPublicImports(const char* name) const {
...
 
   for_vector_allowing_0s(VisibilityStmt, visStmt, mUseImportList) {
-    if (ImportStmt *is = toImportStmt(visStmt)) {
-      if (!is->isPrivate) {
-        if (Symbol *importSym = visStmt->checkIfModuleNameMatches(name)) {
-          if (isModuleSymbol(importSym)) {
-            retval = importSym;
+    if (visStmt != NULL) {
```
Comment (e-kayrakli) [inline compiler/passes/ResolveScope.cpp:983]: I don't know (can't remember?) how we could have NULLs in `mUseImportList`. We clearly can have and that's why we (I?) used that loop there. But the fact that there can be NULLs there seems a bit unfortunate.

Clearly, not part of this PR, thought.

File: compiler/passes/ResolveScope.cpp [hunk 3]
```diff
@@ -1001,12 +1001,14 @@ Symbol* ResolveScope::lookupPublicUnqualAccessSyms(const char* name,
...
-  Symbol *retval = lookupPublicUnqualAccessSyms(name, context, renameLocs);
+  Symbol *retval = lookupPublicUnqualAccessSyms(name, context, renameLocs,
+                                                true);
   return retval;
 }
 
 Symbol* ResolveScope::lookupPublicUnqualAccessSyms(const char* name,
-         BaseAST *context, std::map<Symbol*, astlocT*>& renameLocs) {
+         BaseAST *context, std::map<Symbol*, astlocT*>& renameLocs,
+                                                   bool followUses) {
```
Comment (e-kayrakli) [inline compiler/passes/ResolveScope.cpp:1011]: I prefer this arg to be aligned with `BaseAST` in the line above, can ignore if you disagree

File: compiler/passes/ResolveScope.cpp [hunk 4]
```diff
@@ -1015,22 +1017,28 @@ Symbol* ResolveScope::lookupPublicUnqualAccessSyms(const char* name,
   bool hasPublicImport = false;
```
Comment (e-kayrakli) [inline compiler/passes/ResolveScope.cpp:1017]: I prefer this be renamed to cover UseStmts, as well

File: compiler/passes/ResolveScope.cpp (hunk 2 above)
Comment (lydia-duncan) [reply compiler/passes/ResolveScope.cpp:983]: It's a way to separate the various distances for the use statements - they're traversed in a breadth-first manner, and we only want symbols at one distance to potentially conflict with symbols at the same distance, rather than conflicting with all symbols visible due to use or import statements

Comment (e-kayrakli) [reply compiler/passes/ResolveScope.cpp:983]: Ah yeah, now I remember. Still, I don't like that we have to write this check. But don't have a quick-to-implement solution for that.

File: compiler/passes/ResolveScope.cpp (hunk 3 above)
Comment (lydia-duncan) [reply compiler/passes/ResolveScope.cpp:1011]: I agree, it just was a pain because really the BaseAST should be aligned with the const char above it, but the map argument then wraps around.  I'll play with it, I think there's another order that doesn't exceed 80 characters

Comment (e-kayrakli) [reply compiler/passes/ResolveScope.cpp:1011]: You can typedef that map's type? function name isn't helping either, though I don't want it to be less self-explanatory

Comment (lydia-duncan) [reply compiler/passes/ResolveScope.cpp:1011]: In the past, I've moved the return type up a line and that usually is sufficient, though it may not be in this case.

File: compiler/passes/ResolveScope.cpp (hunk 2 above)
Comment (lydia-duncan) [reply compiler/passes/ResolveScope.cpp:983]: Yeah . . . maybe like a vector of vectors, where index i corresponds to distance i?  I'll open a low priority refactor issue

Comment (e-kayrakli) [reply compiler/passes/ResolveScope.cpp:983]: Right. May be obvious but it'd be great if that refactor adds a macro to iterate such vector of vectors in the needed order.

Comment (lydia-duncan) [reply compiler/passes/ResolveScope.cpp:983]: I opened #15858
